pkg/plugins/policies/meshtimeout/plugin/v1alpha1/plugin.go

```diff
@@ -25,6 +37,216 @@ func (p plugin) MatchedPolicies(dataplane *core_mesh.DataplaneResource, resource
+			for _, dest := range destinations {
+				clusterName, err := gateway_route.DestinationClusterName(dest, hostInfo.Host.Tags)
+				if err != nil {
+					continue
+				}
+				cluster, ok := gatewayClusters[clusterName]
+				if !ok {
+					continue
+				}
+
+				routeActions, ok := routeActionsPerCluster[clusterName]
+				if !ok {
+					continue
+				}
+
+				serviceName := getServiceNameFromListenerInfo(listenerInfo, dest)
+				if err := configure(
+					toRules.Rules,
+					xds.MeshService(serviceName),
+					toProtocol(listenerInfo.Listener.Protocol),
+					gatewayListener,
+					cluster,
+					routeActions,
+				); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func configure(rules core_xds.Rules, subset core_xds.Subset, protocol core_mesh.Protocol, listener *envoy_listener.Listener, cluster *envoy_cluster.Cluster, routeActions []*envoy_route.RouteAction) error {
+	var conf api.Conf
+	if computed := rules.Compute(subset); computed != nil {
+		conf = computed.Conf.(api.Conf)
+	} else {
+		return nil
+	}
+
+	configurer := plugin_xds.Configurer{
+		Conf:     conf,
+		Protocol: protocol,
+	}
+
+	for _, chain := range listener.FilterChains {
+		if err := configurer.ConfigureListener(chain); err != nil {
+			return err
+		}
+	}
+
+	for _, routeAction := range routeActions {
+		configurer.ConfigureRouteAction(routeAction)
+	}
+
+	if err := configurer.ConfigureCluster(cluster); err != nil {
+		return err
+	}
 	return nil
 }
+
+func inferProtocol(routing core_xds.Routing, serviceName string) core_mesh.Protocol {
+	var allEndpoints []core_xds.Endpoint
+	endpoints := core_xds.EndpointList(routing.OutboundTargets[serviceName])
+	allEndpoints = append(allEndpoints, endpoints...)
+	endpoints = routing.ExternalServiceOutboundTargets[serviceName]
+	allEndpoints = append(allEndpoints, endpoints...)
+
+	return envoy_common.InferServiceProtocol(allEndpoints)
+}
+
+func toProtocol(p mesh_proto.MeshGateway_Listener_Protocol) core_mesh.Protocol {
+	return core_mesh.ParseProtocol(mesh_proto.MeshGateway_Listener_Protocol_name[int32(p.Number())])
+}
+
+func routeActionPerCluster(route *envoy_route.RouteConfiguration) map[string][]*envoy_route.RouteAction {
+	actions := map[string][]*envoy_route.RouteAction{}
+	for _, vh := range route.VirtualHosts {
+		for _, r := range vh.Routes {
+			routeAction := r.GetRoute()
+			if routeAction == nil {
+				continue
+			}
+			cluster := routeAction.GetWeightedClusters().GetClusters()[0].Name
+			if actions[cluster] == nil {
+				actions[cluster] = []*envoy_route.RouteAction{routeAction}
+			} else {
+				actions[cluster] = append(actions[cluster], routeAction)
+			}
+		}
+	}
+	return actions
+}
+
+func getServiceNameFromListenerInfo(listenerInfo gateway_plugin.GatewayListenerInfo, dest *gateway_route.Destination) string {
+	endpoints := listenerInfo.OutboundEndpoints[dest.Destination[mesh_proto.ServiceTag]]
+	if len(endpoints) >= 1 {
+		return endpoints[0].Tags["app"]
```

Comment:
how does it work? Is `app` tag always added?

Reply:
if you've checked it and it's true please take it out to a constant because grepping the codebase for "app" to find all the places that write to it would be hard